packages/firefly/src/AppRouter.tsx

```diff
@@ -0,0 +1,188 @@
+import { useIsMswStarted } from "@squide/msw";
+import { useIsRouteMatchProtected, useLogger, useRoutes, type Route } from "@squide/react-router";
+import { useAreModulesReady, useAreModulesRegistered } from "@squide/webpack-module-federation";
+import { useEffect, useMemo, useState, type ReactNode } from "react";
+import { Outlet, RouterProvider, createBrowserRouter, useLocation, type RouterProviderProps } from "react-router-dom";
+
+export type OnLoadPublicDataFunction = (signal: AbortSignal) => Promise<unknown>;
+
+export type onLoadProtectedDataFunction = (signal: AbortSignal) => Promise<unknown>;
+
+export type OnCompleteRegistrationFunction = () => Promise<unknown>;
+
+interface RootRouteProps {
+    fallback: ReactNode;
+    onLoadPublicData?: OnLoadPublicDataFunction;
+    onLoadProtectedData?: onLoadProtectedDataFunction;
+    onCompleteRegistration?: OnCompleteRegistrationFunction;
+    waitForMsw: boolean;
+    areModulesRegistered: boolean;
+    areModulesReady: boolean;
+}
+
+// Most of the bootstrapping logic has been moved to this component because AppRouter
+// cannot leverage "useLocation" since it must be used in a child component of "RouterProvider".
+export function RootRoute(props: RootRouteProps) {
+    const {
+        fallback,
+        onLoadPublicData,
+        onLoadProtectedData,
+        onCompleteRegistration,
+        waitForMsw,
+        areModulesRegistered,
+        areModulesReady
+    } = props;
+
+    // Initialize as loaded if no handler is provided to load the public data.
+    const [isPublicDataLoaded, setIsPublicDataLoaded] = useState(!onLoadPublicData);
+
+    // Initialize as loaded if no handler is provided to load the protected data.
+    const [isProtectedDataLoaded, setIsProtectedDataLoaded] = useState(!onLoadProtectedData);
+
+    const logger = useLogger();
+    const location = useLocation();
+
+    // Re-render the app once MSW is started, otherwise, the API calls for module routes will return a 404 status.
+    const isMswStarted = useIsMswStarted(waitForMsw);
+
+    useEffect(() => {
+        // Only log these messages if MSW is enabled.
+        if (waitForMsw) {
+            if ((areModulesRegistered || areModulesReady) && !isMswStarted) {
+                logger.debug(`[shell] Modules are ${areModulesReady ? "ready" : "registered"}, waiting for MSW to start...`);
```

Comment:
is `shell` still the right word here?

Reply:
Wasn't sure! I initially used "firefly" but since the consumer shell will probably wrap Firefly's `<AppRouter>` component, I thought it was better to use "shell" since the consumer code could also use "shell" and everything would be homogenous instead of seeing likes for [firefly] and [shell] where the person whose reading the logs doesn't care about that distinction and it's only confusing.

Does it make sense?